### PR TITLE
Emit superclass - needs testing and review

### PR DESCRIPTION
**Pull request by bitsol**
The magic sauce:
https://github.com/json-helpers/quicktype/blob/emit-superclass/src/Language/CSharp.ts#L271

### CODE REVIEW

**Comment by schani** [comment]
There are compile errors here.  Before submitting a PR, please make sure that it at least compiles without errors on your machine with `npm run build`.

File: src/Language/CSharp.ts [hunk 1]
```diff
@@ -243,12 +243,15 @@ export class CSharpRenderer extends ConvenienceRenderer {
             return csType;
         }
     }
-
+    protected superclassForClass(): Sourcelike | undefined {
```
Comment (schani) [inline src/Language/CSharp.ts]: This should take a `ClassType` as an argument (that's why it's called `xxxForClass`).  In fact, come to think of it, you might want superclasses for non-class types, too, like for unions.  So let's instead make it

```csharp
    protected superclassForType(_t: Type): Sourcelike | undefined
```

File: src/Language/CSharp.ts [hunk 2]
```diff
@@ -280,51 +287,58 @@ export class CSharpRenderer extends ConvenienceRenderer {
...
-                    this.emitDescription(description);
-                    this.emitLine(property);
-                } else if (this.dense && attribute !== undefined) {
-                    const comment = description === undefined ? "" : ` // ${description.join("; ")}`;
-                    columns.push([attribute, " ", property, comment]);
-                } else {
-                    this.emitDescription(description);
-                    if (attribute !== undefined) {
-                        this.emitLine(attribute);
+        this.emitType(
```
Comment (schani) [inline src/Language/CSharp.ts]: Any chance you could make this change without having to reformat the whole method? It makes it really hard to figure out what's going on in this commit if it contains changes that just reformat code.

File: src/Language/CSharp.ts [hunk 3]
```diff
@@ -280,51 +287,58 @@ export class CSharpRenderer extends ConvenienceRenderer {
...
+            this.descriptionForType(c),
+            AccessModifier.Public,
+            "partial class",
+            className,
+            this.superclassForClass(),
```
Comment (schani) [inline src/Language/CSharp.ts]: `this.superclassForType(c)`

File: src/Language/CSharp.ts [hunk 4]
```diff
@@ -280,51 +287,58 @@ export class CSharpRenderer extends ConvenienceRenderer {
...
+                }
             }
-        });
+        );
     }
 
     private emitUnionDefinition(u: UnionType, unionName: Name): void {
         const nonNulls = removeNullFromUnion(u)[1];
-        this.emitType(this.descriptionForType(u), AccessModifier.Public, "partial struct", unionName, () => {
+        this.emitType(this.descriptionForType(u), AccessModifier.Public, "partial struct", unionName, undefined, () => {
```
Comment (schani) [inline src/Language/CSharp.ts]: Now here we can use `this.superclassForType(u)`

File: src/Language/CSharp.ts (hunk 2 above)
Comment (schani) [reply src/Language/CSharp.ts]: We're using [prettier](https://github.com/prettier/prettier) to format TypeScript.

File: src/Language/CSharp.ts [hunk 5]
```diff
@@ -522,7 +536,7 @@ export class NewtonsoftCSharpRenderer extends CSharpRenderer {
             typeKind = "class";
         }
         const csType = this.csType(t);
-        this.emitType(undefined, AccessModifier.Public, [partial, typeKind], name, () => {
+        this.emitType(undefined, AccessModifier.Public, [partial, typeKind], name, this.superclassForClass(), () => {
```
Comment (schani) [inline src/Language/CSharp.ts]: `this.superclassForType(t)`

File: src/Language/CSharp.ts [hunk 6]
```diff
@@ -537,6 +551,7 @@ export class NewtonsoftCSharpRenderer extends CSharpRenderer {
             AccessModifier.None,
             "static class",
             defined(this._enumExtensionsNames.get(enumName)),
+            this.superclassForClass(),
```
Comment (schani) [inline src/Language/CSharp.ts]: `this.superclassForType(e)`

File: src/Language/CSharp.ts [hunk 7]
```diff
@@ -699,7 +714,7 @@ export class NewtonsoftCSharpRenderer extends CSharpRenderer {
 
     private emitSerializeClass(): void {
         // FIXME: Make Serialize a Named
-        this.emitType(undefined, AccessModifier.Public, "static class", "Serialize", () => {
+        this.emitType(undefined, AccessModifier.Public, "static class", "Serialize", this.superclassForClass(), () => {
```
Comment (schani) [inline src/Language/CSharp.ts]: This class does not correspond to any JSON type - it doesn't need a superclass, so just pass `undefined`.

File: src/Language/CSharp.ts [hunk 8]
```diff
@@ -755,7 +770,7 @@ export class NewtonsoftCSharpRenderer extends CSharpRenderer {
         // FIXME: Make Converter a Named
         let converterName: Sourcelike = ["Converter"];
         if (jsonConverter) converterName = converterName.concat([": JsonConverter"]);
-        this.emitType(undefined, AccessModifier.Internal, "class", converterName, () => {
+        this.emitType(undefined, AccessModifier.Internal, "class", converterName, this.superclassForClass(), () => {
```
Comment (schani) [inline src/Language/CSharp.ts]: Same here I think.

File: src/Language/CSharp.ts (hunk 1 above)
Comment (bitsol) [reply src/Language/CSharp.ts]: I'll switch the method around, then.

Comment (schani) [reply src/Language/CSharp.ts]: Switch it around?  I'm not sure what that means.

Comment (bitsol) [reply src/Language/CSharp.ts]: Sorry. I meant change the arguments and name.

**Pull request merged**